### PR TITLE
8257037: No javac warning when calling deprecated constructor with diamond

### PR DESCRIPTION
**Pull request by lgxbslgx**
Hi all,

When calling deprecated constructor with diamond, the compiler doesn't output warning.
The test case is shown below.

```
GenericClass<Object> o2 = new GenericClass<>();

public class GenericClass<T> {
    @Deprecated
    public GenericClass() {}
}
```

This patch solves the bug and adds corresponding test case.
Thank you for taking the time to review.

Best Regards.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8257037](https://bugs.openjdk.java.net/browse/JDK-8257037): No javac warning when calling deprecated constructor with diamond

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1490/head:pull/1490`
`$ git checkout pull/1490`

### CODE REVIEW

**Comment by lgxbslgx** [comment]
Sorry for closing the PR accidentally.

**Review by mcimadamore — COMMENTED** [review]
I think the fix seems in spirit with respect to what happens for ordinary constructor resolution. I guess there is a question as to whether we should check for these warnings during resolution or after (probably checking for warning after resolution might be a better idea) - as there is some duplication between the code in `findConstructor` and the new `findDiamond` and what goes on in `Attr::checkId` which also check for these warnings (and some more) but doesn't apply checks on constructors.

Just for fun, I tried this patch:

```
diff --git a/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java b/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java
index df21524229c..05d56136000 100644
--- a/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java
+++ b/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java
@@ -4459,15 +4459,11 @@ public class Attr extends JCTree.Visitor {
             }
 
             // Emit a `deprecation' warning if symbol is deprecated.
-            // (for constructors (but not for constructor references), the error
-            // was given when the constructor was resolved)
 
-            if (sym.name != names.init || tree.hasTag(REFERENCE)) {
-                chk.checkDeprecated(tree.pos(), env.info.scope.owner, sym);
-                chk.checkSunAPI(tree.pos(), sym);
-                chk.checkProfile(tree.pos(), sym);
-                chk.checkPreview(tree.pos(), sym);
-            }
+            chk.checkDeprecated(tree.pos(), env.info.scope.owner, sym);
+            chk.checkSunAPI(tree.pos(), sym);
+            chk.checkProfile(tree.pos(), sym);
+            chk.checkPreview(tree.pos(), sym);
 
             // If symbol is a variable, check that its type and
             // kind are compatible with the prototype and protokind.
diff --git a/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java b/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java
index 727d2989adf..9bb37fc23b8 100644
--- a/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java
+++ b/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java
@@ -2859,8 +2859,6 @@ public class Resolve {
                                     names.init, argtypes,
                                     typeargtypes, allowBoxing,
                                     useVarargs);
-        chk.checkDeprecated(pos, env.info.scope.owner, sym);
-        chk.checkPreview(pos, sym);
         return sym;
     }
 
@@ -2937,6 +2935,7 @@ public class Resolve {
                             return sym;
                         }
                     };
+                    newConstr.setAttributes(sym);
                     bestSoFar = selectBest(env, site, argtypes, typeargtypes,
                             newConstr,
                             bestSoFar,
```
This fixes your issues, and removes the warning code duplication between `Attr` and `Resolve`, as well as adding checks for all the remaining kinds of warnings (sunAPI, deprecation, preview and profile) - although I believe most of those are not as important.

I guess I'll leave you to decide how you want to tackle this. As I said, the patch you wrote remains within the spirit of the original code - my only minor quibble is that this "spirit" is questionable, as it involves issuing same warnings in different places, this enforcing developers to replicate code (and forgetting to do so, as it already happened) in multiple places.

**Comment by lgxbslgx** [comment]
Thank you for the comment and suggestion.

I had seen and paid attention to the following code in `Attr.java` when being solved this bug.

```
            // (for constructors (but not for constructor references), the error
            // was given when the constructor was resolved)

            if (sym.name != names.init || tree.hasTag(REFERENCE)) {
                chk.checkDeprecated(tree.pos(), env.info.scope.owner, sym);
                chk.checkSunAPI(tree.pos(), sym);
                chk.checkProfile(tree.pos(), sym);
                chk.checkPreview(tree.pos(), sym);
            }
```

These code confused me. Because the comments `(for constructors, the error was given when the constructor was resolved)`  and the conditional expression `sym.name != names.init` were submitted at 2007(submit log: Initial load). And nobody raises suggestion to it over these years though this code snippet had been revised many times.

In order not to generate unknown issues, I chose not to revise these code when being solved this bug.

@mcimadamore  I test your patch locally by using the following command. All the tests passed.

```
make test TEST="jtreg:langtools/tools/javac" CONF=linux-x86_64-server-release
```

Your patch is a better solution if no any unknown regression occurs. But I have no idea about the unknown regressions now.
I would like to accept your suggestion and revise my code. **And we should think more situations which could generate regressions to avoid unknown issues occur.**

----
Finally, I have a little question.
Is the code `newConstr.setAttributes(sym);` necessary in `Resolve.findDiamond`?
I delete it in your patch, and test locally by using  the following command. All the tests passed.

```
make test TEST="jtreg:langtools/tools/javac" CONF=linux-x86_64-server-release
```

**Comment by mcimadamore** [comment]
>  Is the code `newConstr.setAttributes(sym);` necessary in `Resolve.findDiamond`?

Sorry, that was a leftover from a previous experiment. That's not necessary. We'll run some more tests on our side and I'll let you know.

**Comment by lgxbslgx** [comment]
@mcimadamore Should I add you as a contributor? Does it need another reviewer to review if I add you as a contributor?

**Comment by mcimadamore** [comment]
> @mcimadamore Should I add you as a contributor of this patch? Does it need another reviewer to review if I add you as a contributor?

As you prefer - I don't mind either way. I don't think adding me as a contributor should prevent the integration, or me being able to sponsor, but I'm not 100% sure.

**Comment by lgxbslgx** [comment]
I revise the test code based on the comment[1] written by Jonathan Gibbons in JDK-8231622[2].

> For tests that use /ref=file -XDrawDiagnostics the conventions are:
>
>   1. Omit the complete legal header, including the copyright and license
>   2. After `@test` add the text `/nodynamiccopyright/`
>
> #1 protects the file against any future changes in the length of the legal header, that might affect line numbers, and #2 is for use by automated scripts that may check for the presence of the legal header.

[1] https://mail.openjdk.java.net/pipermail/compiler-dev/2020-December/015542.html
[2] https://github.com/openjdk/jdk/pull/1626

**Comment by mcimadamore** [comment]
The extensive testing revealed at least 10 cases where the compiler is issuing a new, resolution-related error when it wasn't doing so before - (the errors seem to have to do with `abstract` methods). As a result I think it's best to go with your first conservative approach, to avoid troubles. I'll try to investigate further and come up with a reduced test case for the failing condition - but that doesn't have to hold up this fix. Thanks!

**Comment by lgxbslgx** [comment]
It is a good news that the trouble is found before the integration instead of after the release of JDK16.
I revise the code to previous version.
Thank your for the review.

/contributor remote @mcimadamore

**Comment by mcimadamore** [comment]
Actually, after some more analysis, I discovered that many of the errors were unrelated with your fix - and had to do instead with the addition of a new default method on Stream, which is causing overload resolution changes in code that has static imports on Collectors.

Only one issue remains to be fully triaged - but, being annotation related, I think it's unlikely caused by your patch. In other words, I believe both patches are likely to be safe. I'm obviously ok with you going ahead with the more conservative fix, since you already reverted the code to that version.

I apologize for the back and forth!

**Review by mcimadamore — APPROVED** [review]
Looks good to go

**Comment by mcimadamore** [comment]
>  you mean Collectors.toList() vs Stream.toList()

yes - precisely

**Pull request closed without merging**
